### DISCUSSION TITLE
How to get the volume of mesh?

### DISCUSSION THREAD

**Discussion by hhy2022** [idaholab/moose#21506] [Q&A General] [answered]
Hi,

I am trying to output the volume information about the mesh I used. I found some useful example in [https://mooseframework.inl.gov/source/auxkernels/VolumeAux.html](url). 

I run the example and got the result below:
This was the mesh of the problem:
![1](https://user-images.githubusercontent.com/100750125/177387198-538059d9-1cf2-49fa-b974-afc536483a64.png)
This was the volume output of the problem:
![2](https://user-images.githubusercontent.com/100750125/177387278-50d7d178-fc26-418d-8024-cb389c1f0ff6.png)

I am a little confused since I thought the output volume didn't match the real situation. 

I went to the source code and found the volume was calculated using `return _bnd ? _current_side_volume : _current_elem_volume;`. I thought it should calculate the volume of the mesh. I might misunderstood something and perhaps there were some problems with my output method. I am wondering if anyone can get me some suggestions on getting the volume of the mesh. I should use this value as a parameter in developing my own application. 

Thank you so much for your help.

**Comment by GiudGiud** [comment, marked as answer]
Hello

So you want the volume of the mesh, not each cell in the mesh?
if so, you may use this postprocessor
https://mooseframework.inl.gov/source/postprocessors/VolumePostprocessor.html

Guillaume

**Reply by hhy2022** [reply]
I want each cell volume. Sorry about the mistake.

Huihua

**Reply by GiudGiud** [reply]
then VolumeAux is the way. Is the result wrong? What is the problem?

**Reply by hhy2022** [reply]
I thought the result should provide 20 volume values since there were 20 cells. But the output only showed 6 different color bars and the values didn't match the manually calculated ones.

For example, the left down corner cell should have a volume around 0.4875 based on the mesh information calculation (1.5/2*1.3/2=0.4875). However, the light blue (left down corner cell) showed a value around 0.4125.

**Reply by GiudGiud** [reply]
ah yes that is weird indeed!
Could you please paste your input file syntax for the VolumeAux ?

**Reply by hhy2022** [reply]
Sure. Actually it was an example shown in the link [https://mooseframework.inl.gov/source/auxkernels/VolumeAux.html](url).
Below is the detail:
```
[Mesh]
  [cmg]
    type = CartesianMeshGenerator
    dim = 2
    dx = '1.5 2.4 0.1'
    dy = '1.3 0.9'
    ix = '2 1 1'
    iy = '2 3'
  []
[]

[AuxVariables/volume]
  order = CONSTANT
  family = MONOMIAL
[]

[AuxKernels/volume_aux]
  type = VolumeAux
  variable = volume
[]

[Problem]
  solve = false
[]

[Executioner]
  type = Steady
[]

[Outputs]
  exodus = true
  csv = true
[]
```

**Reply by hhy2022** [reply]
The total output showed 5 different color bars and that were correct since there were 5 different volumes for this problem. Sorry about the misunderstood before.

The only porblem now was the exact value of the volume seemed didn't match the manually calculated ones.

**Reply by GiudGiud** [reply]
It matches on my machine?
For some of the areas for example, I get
0.4875 in the bottom left, 1.56 in the bottom middle/right and 0.03 for the top right. 

now analytically we should have:
1.5/2 * 1.3/2 = 0.4875
2.4/1 * 1.3/2 = 1.56
0.1/1 * 0.9/3 = 0.03

**Reply by hhy2022** [reply]
Yes. But the color bar value showing that the bottom left should be 0.4125 while the calculated one was 0.4875. The green one from color bar was 0.795 while the calculated one should be 0.72.
Wait, is there any different with the color? like light green and dark green? I thought the color bar with values on side should be the indicating values of the volume. Can I use another output method to print out the real value instead of the color bar figure?

**Reply by GiudGiud** [reply]
in paraview you may hover over the cells and get the values
I dont know about peacock

**Reply by hhy2022** [reply]
So this VolumeAux can get the cell volume well. Right? 
Besides, I can use paraview, but I couldn't see the values when I hover the cells. Could you please tell me which version you used?

**Reply by GiudGiud** [reply]
you have to click on a button on the top bar to enable that

**Reply by hhy2022** [reply]
I see. Thank you so much for your help.

**Comment by loganharbour** [comment]
You can also run mesh only mode (`--mesh-only` on command line) to get some output like this:

```
<cut>

  Mesh Subdomains:
   Subdomain 0: 16 elems (QUAD4, 16 active), 25 active nodes
    Volume: 1
    Bounding box minimum: (x,y,z)=(       0,        0,        0)
    Bounding box maximum: (x,y,z)=(       1,        1,        0)
    Bounding box delta: (x,y,z)=(       1,        1,        0)
   Global mesh volume = 1
```

where it will show the volume of each subdomain and the total volume.

**Reply by hhy2022** [reply]
OK. Thank you so much for that information. I will have a try.
